Support FOpts encryption usage errata

adriansmares opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes https://github.com/TheThingsNetwork/lorawan-stack/issues/5360

#### Changes
<!-- What are the changes made in this pull request? -->

- Remove support for LoRaWAN version comparisons in the codebase
  - These were hard to maintain, since we had lots of checks spread throughout the Network Server and Join Server
  - They have been replaced by the `macspec` package, which based on a LoRaWAN version can decide if certain behavior is available
- Support the FOpts encryption erratum
  - This one is achieved by introducing the concept of encryption options, which manipulate how data is encrypted
  - The `macspec` package can generate encryption options based on the LoRaWAN version
- Avoid starting the downlink task dispatcher in some tests
  - Fixes some annoying flaky tests that had the `Dispatch called, but not set` error


#### Testing

<!-- How did you verify that this change works? -->

Tested with the 4.6.0 release of LoRaMac-node. I've tested basically all 3 situations:
- FOpts attached to network (FPort = 0) downlink - start with ADR margin 60 dBm, allow device to send some uplinks in SF12. Change margin to 15 dBm. This forces a `LinkADRReq` which will be part of the FOpts. The corresponding `LinkADRAns` confirms that the `LinkADRReq` was received.
- FOpts attached to application (FPort != 0) downlink - same instructions, but also add an application downlink before you change the margin to 15 dBm.
- FOpts attached to uplink - both of the above tests confirm that this works, as `LinkADRAns` arrives for both in the FOpts field if the end device always sends data uplinks.

##### Regressions

<!-- Please indicate features that this change could affect and how that was tested. -->

I could not find a LoRaMac-node version that implements the pre-erratum version. I don't know if these devices actually exist in the wild.

#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

adriansmares committed:
all: Refactor MAC version comparisons

adriansmares committed:
ns: Do not start downlink task dispatcher

adriansmares commented on pkg/networkserver/downlink.go:
```diff
@@ -229,60 +230,52 @@ func (ns *NetworkServer) generateDataDownlink(ctx context.Context, dev *ttnpb.En
-							MaxDownLen: maxDownLen,
-							MaxUpLen:   maxUpLen,
-						}
+		enqueuers = append(enqueuers,
```
I've removed the version checks here since each individual  `mac.Enqueue*` command does the version checks as well. I can bring them back, but I'm not sure how much value that would bring (especially since arithmetic comparisons are no longer what we should do).

johanstokking commented:
Conceptually this `macspec` is a fantastic idea.

Regarding 1.1 vs 1.1 + errata: the original 1.1 is considered broken and can only work correctly with the errata. So this would rather be a fix for 1.1; no need for another version. Right?

I'll review tomorrow.

adriansmares commented:
> Regarding 1.1 vs 1.1 + errata: the original 1.1 is considered broken and can only work correctly with the errata. So this would rather be a fix for 1.1; no need for another version. Right?

I actually don't know. I've read this statement, that 1.1 was 'not possible' without the errata, but I don't really understand how development actually occurred since 2018 related to 1.1 - I would have expected to receive more bug reports regarding this in the last 2 years. 

We can definitely just add this behavior to the 1.1 version, but this is a breaking change and I would like to understand better if we really have and will have 0 original 1.1 end devices.

For example https://github.com/TheThingsNetwork/lorawan-stack/pull/2339 came in April 2020, and it basically introduces the 'vanilla' 1.1 encryption behavior. Did we have collective amnesia with respect to the errata (which came in January 2018) at the time ? Because both LoRaMac-node and mbed-OS did not use the errata at the time, and apparently people were testing MIC checks and MAC commands 🤔 .

htdvisser reviewed (approved):
Code-owned files LGTM.

Package `macspec` seems like a good idea. Please make sure to add it to `CODEOWNERS`.

johanstokking commented:
> I actually don't know. I've read this statement, that 1.1 was 'not possible' without the errata, but I don't really understand how development actually occurred since 2018 related to 1.1 - I would have expected to receive more bug reports regarding this in the last 2 years.
> 
> We can definitely just add this behavior to the 1.1 version, but this is a breaking change and I would like to understand better if we really have and will have 0 original 1.1 end devices.

I don't think anyone uses LoRaWAN 1.1 seriously, especially not the non-errata.

This is the timeline:

- October 2017: LoRaWAN 1.1
- January 2018: LoRaWAN 1.1 FCnt errata (with Christmas and New Year in between, plus TC meeting schedule, this is effectively an immediate fix)
- July 2018: LoRaWAN 1.0.3
- October 2020: LoRaWAN 1.0.4

It was pretty clear in 2018 that LoRaWAN 1.1 was too far fetched; the device maker community wanted a simpler stack with minor additions/fixes compared to 1.0.2 and more regions. So from what I know, nothing happened to LoRaWAN 1.1 since the spring of 2018 and all effort went in LoRaWAN 1.0.3, 1.0.4 and RP2.

> For example #2339 came in April 2020, and it basically introduces the 'vanilla' 1.1 encryption behavior. Did we have collective amnesia with respect to the errata (which came in January 2018) at the time ?

Yes, I think so.

> Because both LoRaMac-node and mbed-OS did not use the errata at the time, and apparently people were testing MIC checks and MAC commands 🤔 .

Early LoRaWAN 1.1 implementations have mostly been implementations without actual interoperability testing on actual LoRaWAN networks. I confirm this specifically for the Mbed OS stack.

I think we should consider erratas really as fixes. Otherwise there would have been a 1.1.1 version instead.

adriansmares commented:
Per above conversation I've removed the `1_1_FCNT` version and incorporated the changes into the currently existing `1_1` version. Feel free to review @johanstokking .

johanstokking commented on pkg/crypto/data_messages.go:
```diff
@@ -24,7 +24,26 @@ import (
+// EncryptionOption encapsulates custom LoRaWAN encryption options.
+type EncryptionOption func(*encryptionOptions)
+
+// WithFrameIdentifier specifies which bytes are to be used for
+// positions [1:5] during encryption.
+func WithFrameIdentifier(identifier [4]byte) EncryptionOption {
+	return func(encOpts *encryptionOptions) {
+		copy(encOpts.frameIdentifier[:], identifier[:])
+	}
+}
```
Can you document this a bit more? What is the LoRaWAN aspect of this?

adriansmares replied:
This is a bit of poetic license from my side. The (updated) specification does not have names for the parts of the block:
![image](https://user-images.githubusercontent.com/36161392/169006956-2f274779-6b1c-4d16-8dcc-ee5f56193860.png)
The 4 bytes after the `0x01` are what I call the frame identifier. The naming itself comes from the erratum body:
![image](https://user-images.githubusercontent.com/36161392/169007328-0567bdcb-6f25-458e-98a5-4e14b04c9530.png)
They are called 'constants' there, but their role is to separate between different frame types - so maybe 'frame type constant' is closer to what they really do ?

johanstokking replied:
Yes that sounds better to me.

johanstokking commented on pkg/specification/macspec/specification.go:
```diff
@@ -0,0 +1,242 @@
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) >= 0
+}
+
+// UseLegacyMIC reports whether v uses legacy MIC computation algorithm.
+func UseLegacyMIC(v ttnpb.MACVersion) bool {
```
Legacy is a bit of an unfortunate name in The Things Stack. Initially, in 2018 when we started with The Things Stack, LoRaWAN 1.1 was the future and so LoRaWAN 1.0.x was legacy. That didn't really turn out to be the case.

With LoRaWAN 1.2, the MIC is going to change again. I think therefore that these shouldn't be boolean switches but rather per-version functions that compute the MIC. If we do that in `pkg/crypto`, do we need this function here?

This is not really a need for change now, but some foreseeable technical debt with LoRaWAN 1.2 on the horizon.

adriansmares replied:
We can have the equivalent of `crypto.EncryptionOption` for MIC - `crypto.MICOption` that basically controls how the MIC is generated, and then the `macspec` package can choose based on the LoRaWAN version what options to provide.

What is a bit different is that the LoRaWAN 1.1 MIC is basically derived from the 1.0.x MIC, so for now we're always computing the 1.0.x MIC. I guess we will see when 1.2 is out what kind of changes to introduce.

johanstokking replied:
The 1.1 MIC is already different, right, taking two integrity keys, ConfFCnt and TX parameters into account?

What will happen with 1.2 is that there comes a companion spec that specifies cipher suites. They will declare the MIC input and keys to use. Indeed let's see how that comes out of the washing machine and dryer and not address anything here and now.

adriansmares replied:
It is different but it acts as an extension of the 1.0.x MIC. So each time when you compute the 1.1 MIC, your first step is to actually compute the 'legacy' mic (which is 2 bytes), then you compute the 1.1 extension from there (the other two bytes).

This is an optimization used by the NS right now:
- You always compute the expected legacy MIC
- You always compare the expected legacy MIC with the provided legacy MIC. If they mismatch, you can already drop the frame since extending it with the 1.1 procedure won't make it match.
- If the end device is a 1.1 device, you actually compute the other two bytes and compare the MIC.

johanstokking commented on pkg/specification/macspec/specification.go:
```diff
@@ -0,0 +1,242 @@
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) < 0
+}
+
+// UseSharedFCntDown reports whether v uses the same frame counter
+// for both network and application downlinks.
+func UseSharedFCntDown(v ttnpb.MACVersion) bool {
```
`UseSingleFCntDown` ?

adriansmares replied:
I named it 'shared' since this makes it clear that both entities (application layer and network layer) basically fight/negotiate which value to use. The fact that the frame counter is single could also be understood as only one layer needs to manage it.

johanstokking commented on pkg/specification/macspec/specification.go:
```diff
@@ -0,0 +1,242 @@
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) < 0
+}
+
+// UseDLChannel reports whether v is allowed to use the
+// DLChannel{Req|Ans} MAC commands.
+func UseDLChannel(v ttnpb.MACVersion) bool {
```
```suggestion
// UseDLChannelReq reports whether v is allowed to use the
// DLChannel{Req|Ans} MAC commands.
func UseDLChannelReq(v ttnpb.MACVersion) bool {
```

All versions have downlink channels, but not all versions support the setup with `DLChannelReq`. I found the `UseDLChannelReq(v)` check a bit weirdly looking. This is a bit of a personal preference.

johanstokking commented on pkg/specification/macspec/specification.go:
```diff
@@ -0,0 +1,242 @@
+
+// UseTxParamSetup reports whether v is allowed to use the
+// TxParamSetup{Req|Ans} MAC commands.
+func UseTxParamSetup(v ttnpb.MACVersion) bool {
```
Idem and onwards

johanstokking commented on pkg/specification/macspec/specification.go:
```diff
@@ -0,0 +1,242 @@
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_2) >= 0 && compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) < 0
+}
+
+// AtomicLinkADRAns reports whether v will accept or reject all
+// channel masks in a singular LinkADRAns MAC response.
+func AtomicLinkADRAns(v ttnpb.MACVersion) bool {
```
Then I would also call this `SingularLinkADRAns`

johanstokking commented on pkg/specification/macspec/specification.go:
```diff
@@ -0,0 +1,242 @@
+
+// RequireDevEUIForABP reports whether v requires ABP devices to have a DevEUI associated.
+func RequireDevEUIForABP(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0 && compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) < 0
```
Let's also require this for LoRaWAN 1.1. I'm aware it's not with that many words in the spec, but let's please require this.

johanstokking commented on pkg/specification/macspec/specification.go:
```diff
@@ -0,0 +1,242 @@
+	}
+}
+
+// ValidateUplinkPayloadSize reports whether v requires that the uplink
+// payload size must be valid with respect to the regional parameters
+// and local limitations.
+func ValidateUplinkPayloadSize(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) >= 0
+}
```
Is this a spec change since 1.1?

adriansmares replied:
Full disclosure - I added this because it existed before:
https://github.com/TheThingsNetwork/lorawan-stack/blob/70975ab7e96936103f2e1ca70b53cd68e1a64898/pkg/networkserver/downlink.go#L1825-L1832

The spec itself is quite straight forward about why we do this (page 29):
![image](https://user-images.githubusercontent.com/36161392/169009712-48021fb7-14a5-4ac3-a420-e203d3ec39bb.png)

johanstokking replied:
Okay so this is for downlink. That should be more clear.

johanstokking reviewed (approved):
Looks great, just minor code quality things which you can ignore if you disagree.

adriansmares committed:
console: Require DevEUI for LoRaWAN 1.1 ABP devices

ryaplots reviewed (approved):
code-owned lgtm
